transform.ml

```diff
@@ -78,3 +78,53 @@ let branch_prune (prog : program) : program =
   let final = branch_prune 0 [] in
   let cleanup = remove_empty_jmp (remove_unreachable_code final) in
   ("main", cleanup) :: (deopt_label, main) :: rest
+
+let move instrs from_pc to_pc =
+  let (dir, to_pc) = if from_pc > to_pc then (-1, to_pc) else (1, to_pc-1) in
+  let from = instrs.(from_pc) in
+  let rec move pc =
+    if pc != to_pc then begin
+      instrs.(pc) <- instrs.(pc+dir);
+      move (pc+dir)
+    end
+  in
+  move from_pc;
+  instrs.(to_pc) <- from
+
+let hoist_assignment prog =
```

Comment:
I would also like a comment that explains code motion and its conditions. Something like the text of the PR is good enough.